bint.go

```diff
@@ -210,48 +210,160 @@ func parseBintFromU128(s []byte) (bool, bint, uint8, error) {
 		coef u128
 		prec uint8
 	)
-	for ; pos < width; pos++ {
-		if s[pos] == '.' {
+
+	if len(s[pos:]) <= 19 {
+		coef, prec, err = parseSmallToU128(s[pos:])
+	} else {
+		coef, prec, err = parseLargeToU128(s[pos:])
+	}
+
+	if err == ErrInvalidFormat {
+		return neg, bint{}, 0, errInvalidFormat(s)
+	}
+
+	return neg, bint{u128: coef}, prec, err
+}
+
+func parseSmallToU128(s []byte) (u128, uint8, error) {
+	var (
+		coef uint64
+		prec uint8
+	)
+
+	for i := 0; i < len(s); i++ {
+		if s[i] == '.' {
 			// return err if we encounter the '.' more than once
 			if prec != 0 {
-				return false, bint{}, 0, errInvalidFormat(s)
+				return u128{}, 0, ErrInvalidFormat
 			}
 
 			// nolint: gosec
-			prec = uint8(width - pos - 1)
+			prec = uint8(len(s) - i - 1)
 
 			// prevent "123." or "-123."
 			if prec == 0 {
-				return false, bint{}, 0, errInvalidFormat(s)
+				return u128{}, 0, ErrInvalidFormat
 			}
 
 			if prec > defaultPrec {
-				return false, bint{}, 0, ErrPrecOutOfRange
+				return u128{}, 0, ErrPrecOutOfRange
 			}
 
 			continue
 		}
 
-		if s[pos] < '0' || s[pos] > '9' {
-			return false, bint{}, 0, errInvalidFormat(s)
+		if s[i] < '0' || s[i] > '9' {
+			return u128{}, 0, ErrInvalidFormat
 		}
 
-		coef, err = coef.Mul64(10)
-		if err != nil {
-			return false, bint{}, 0, err
+		coef = coef*10 + uint64(s[i]-'0')
+	}
+
+	if coef == 0 {
+		return u128{}, 0, nil
+	}
+
+	return u128{lo: coef}, prec, nil
+}
+
+func parseLargeToU128(s []byte) (u128, uint8, error) {
+	// find '.' position
+	l := len(s)
+	pos := -1
+
+	for i := 0; i < len(s); i++ {
+		if s[i] == '.' {
+			pos = i
 		}
+	}
+
+	if pos == 0 || pos == l-1 {
+		return u128{}, 0, ErrInvalidFormat
+	}
+
+	var prec uint8
+	if pos != -1 {
+		// nolint: gosec
+		prec = uint8(l - pos - 1)
+		if prec > defaultPrec {
+			return u128{}, 0, ErrPrecOutOfRange
+		}
+	}
 
-		coef, err = coef.Add64(uint64(s[pos] - '0'))
+	if pos == -1 {
```

Comment:
Testing this after a pos != 1 means it could be an else here, no?

Reply:
I didn't really get this. Can you elaborate? What this does is simply handle the case where the input has no decimal point

Reply:
Here is the code once you updated with this PR

```go
        var prec uint8
        if pos != -1 {
                // nolint: gosec
                prec = uint8(l - pos - 1)
                if prec > defaultPrec {
                        return u128{}, 0, ErrPrecOutOfRange
                }
        }

        if pos == -1 {
                // no decimal point
                coef, err := digitToU128(s)
                if err != nil {
                        return u128{}, 0, err
                }

                return coef, 0, nil
        }
```

My comment was to consider using this

```go
        var prec uint8
        if pos != -1 {
                // nolint: gosec
                prec = uint8(l - pos - 1)
                if prec > defaultPrec {
                        return u128{}, 0, ErrPrecOutOfRange
                }
        } else {
                // no decimal point
                coef, err := digitToU128(s)
                if err != nil {
                        return u128{}, 0, err
                }

                return coef, 0, nil
        }
```

Reply:
yep, misunderstood it. I already changed the implementation for this in https://github.com/quagmt/udecimal/pull/18